pkg/usersettings/handlers.go

```diff
@@ -0,0 +1,170 @@
+		return
+	}
+
+	userSettingMeta, err := h.getUserSettingMeta(context, user)
+	if err != nil {
+		klog.Errorf("Failed to get user data to handle user setting request: %v", err)
+		serverutils.SendResponse(w, http.StatusBadGateway, serverutils.ApiError{Err: fmt.Sprintf("Failed to get user data to handle user setting request: %v", err)})
+		return
+	}
+
+	switch r.Method {
+	case http.MethodGet:
+		configMap, err := h.getUserSettings(context, serviceAccountClient, userSettingMeta)
+		if err != nil {
+			klog.Errorf("Failed to get user settings: %v", err)
+			serverutils.SendResponse(w, http.StatusBadGateway, serverutils.ApiError{Err: fmt.Sprintf("Failed to get user settings: %v", err)})
+		} else {
+			serverutils.SendResponse(w, http.StatusOK, configMap)
+		}
+	case http.MethodPost:
+		configMap, err := h.createUserSettings(context, serviceAccountClient, userSettingMeta)
+		if err != nil {
+			klog.Errorf("Failed to create user settings: %v", err)
+			serverutils.SendResponse(w, http.StatusBadGateway, serverutils.ApiError{Err: fmt.Sprintf("Failed to create user settings: %v", err)})
+		} else {
+			serverutils.SendResponse(w, http.StatusOK, configMap)
+		}
+	case http.MethodDelete:
+		err := h.deleteUserSettings(context, serviceAccountClient, userSettingMeta)
+		if err != nil {
+			klog.Errorf("Failed to delete user settings: %v", err)
+			serverutils.SendResponse(w, http.StatusBadGateway, serverutils.ApiError{Err: fmt.Sprintf("Failed to delete user settings: %v", err)})
+		} else {
+			w.WriteHeader(http.StatusNoContent)
+		}
+	default:
+		w.Header().Set("Allow", "GET, POST, DELETE")
+		serverutils.SendResponse(w, http.StatusMethodNotAllowed, serverutils.ApiError{Err: "Unsupported method, supported methods are GET POST DELETE"})
+	}
+}
+
+// Fetch the current user with auth.User token and returning the user-setting ConfigMap.
+func (h *UserSettingsHandler) getUserSettings(ctx context.Context, client *kubernetes.Clientset, userSettingMeta *UserSettingMeta) (*core.ConfigMap, error) {
+	return client.CoreV1().ConfigMaps(namespace).Get(ctx, userSettingMeta.getConfigMapName(), meta.GetOptions{})
+}
+
+// Fetch the current user with auth.User token and create a new user-setting ConfigMap.
+// Returns the existing ConfigMap if it is already exist.
+func (h *UserSettingsHandler) createUserSettings(ctx context.Context, client *kubernetes.Clientset, userSettingMeta *UserSettingMeta) (*core.ConfigMap, error) {
+	role := createRole(userSettingMeta)
+	roleBinding := createRoleBinding(userSettingMeta)
+	configMap := createConfigMap(userSettingMeta)
+
+	_, err := client.RbacV1().Roles(namespace).Create(ctx, role, meta.CreateOptions{})
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return nil, err
+	}
+
+	_, err = client.RbacV1().RoleBindings(namespace).Create(ctx, roleBinding, meta.CreateOptions{})
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return nil, err
```

Comment:
we should call `deleteUserSettings()` method upon error. So we dont end up in an inconsistent state, where Role and RoleBinding was created but the CM for some reason was not.
Either that or we rely that the request would be repeated until all the resources are created.

Reply:
As discussed in slack, I added `deletedUserSettings` before returning the create error.